### PR TITLE
store: cache propose metric and flush in batch.

### PR DESCRIPTION
**Pull request by siddontang**
Like other frequent raft metrics, cache propose metric and flush in batch. 

@ngaut @BusyJay @hhkbp2

### CODE REVIEW

**Comment by siddontang** [comment]
To fix https://github.com/pingcap/tikv/issues/1293

@overvenus

File: src/raftstore/store/local_metrics.rs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+#[derive(Debug, Default, Clone)]
+pub struct RaftReadyMetrics {
+    pub message: u64,
+    pub commit: u64,
+    pub append: u64,
+    pub snapshot: u64,
+}
+
+impl RaftReadyMetrics {
+    /// Flushs all metrics
```
Comment (hhkbp2) [inline src/raftstore/store/local_metrics.rs]: s/Flushs/Flushes/

File: src/raftstore/store/local_metrics.rs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+/// The buffered metrics counters for raft.
+#[derive(Debug, Default, Clone)]
+pub struct RaftMetrics {
+    pub ready: RaftReadyMetrics,
+    pub message: RaftMessageMetrics,
+    pub propose: RaftProposeMetrics,
+}
+
+impl RaftMetrics {
+    /// Flushs all metrics
```
Comment (hhkbp2) [inline src/raftstore/store/local_metrics.rs]: Flushes

File: src/raftstore/store/local_metrics.rs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+pub struct RaftProposeMetrics {
+    pub all: u64,
+    pub local_read: u64,
+    pub normal: u64,
+    pub transfer_leader: u64,
+    pub conf_change: u64,
+}
+
+impl RaftProposeMetrics {
+    /// Flushs all metrics
```
Comment (hhkbp2) [inline src/raftstore/store/local_metrics.rs]: Flushes

**Comment by hhkbp2** [comment]
Rest LGTM

**Comment by siddontang** [comment]
PTAL @ngaut  @BusyJay

File: src/raftstore/store/local_metrics.rs [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            self.commit = 0;
+        }
+        if self.append > 0 {
+            STORE_RAFT_READY_COUNTER_VEC.with_label_values(&["append"])
+                .inc_by(self.append as f64)
+                .unwrap();
+            self.append = 0;
+        }
+        if self.snapshot > 0 {
+            STORE_RAFT_READY_COUNTER_VEC.with_label_values(&["snapshot"]).inc();
```
Comment (disksing) [inline src/raftstore/store/local_metrics.rs]: Only inc by 1?

Comment (disksing) [inline src/raftstore/store/local_metrics.rs]: I think use `HashMap<&str, u64>` will be simpler and more easy to prevent mistakes.

Comment (siddontang) [inline src/raftstore/store/local_metrics.rs]: I prefer field and it is more clear than HashMap.

Comment (siddontang) [inline src/raftstore/store/local_metrics.rs]: @hhkbp2 

Seem we should use inc_by here, not inc.

**Comment by disksing** [comment]
LGTM

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**